qiskit/primitives/containers/observable.py

```diff
@@ -0,0 +1,164 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""
+Sparse container class for an Estimator observable.
+"""
+from __future__ import annotations
+
+from typing import Union, Mapping as MappingType
+from collections.abc import Mapping
+from collections import defaultdict
+from numbers import Complex
+
+from qiskit.quantum_info import Pauli, SparsePauliOp
+
+
+ObservableKey = "tuple[tuple[int, ...], str]"  # pylint: disable = invalid-name
```

Comment:
It might be good to make a dataclass of `ObservableKey`.

Reply:
Im mostly worried about this class being slow and memory intensive with all the nested classes. I did have a prototype using a class that function like a tuple (had getitem), but im a bit skeptical of validation other than via type since it might be slow. We can't validate the strings, so validation would be that qubits is a tuple with no duplicates and (if we want to enforce this) that the string is same lenght as qubits. 

In the current version the design doesn't actually require the string to represent a bunch of operators for each qubit, its up to an estimator to decide what it does with it, but we are using it to represent a Pauli label 

(side note: our reference estimator is going to be rather inefficient since it unpacks SparsePauliOps into this potentially much more memory heavy format, and then has to convert back to SparsePauliOp for its simulation)